Update outdated docs, improve naming, prefer user props over auth hook in docs

FranjoMindek opened the pull request:
- Adds `await` for `sanitizeAndSerializeProviderData` calls
- Renames `data-model/backends` to `data-model/databases` 
- Docs guide users to use `user` props rather than `useAuth()`, also reflected on examples

Related issues: https://github.com/wasp-lang/wasp/issues/2514, https://github.com/wasp-lang/wasp/issues/1071

Martinsos commented on web/docs/advanced/jobs.md:
```diff
@@ -233,7 +233,7 @@ The Job declaration has the following fields:
   Our jobs need job executors to handle the _scheduling, monitoring, and execution_.
 
-  `PgBoss` is currently our only job executor, and is recommended for low-volume production use cases. It requires that your database provider is set to `"postgresql"` in your `schema.prisma` file. Read more about setting the provider [here](../data-model/backends.md#postgresql).
+  `PgBoss` is currently our only job executor, and is recommended for low-volume production use cases. It requires that your database provider is set to `"postgresql"` in your `schema.prisma` file. Read more about setting the provider [here](../data-model/databases.md#postgresql).
```
You learned aboutthis by trying to follow the link?
If so, I wish we knew about this link being wrong without having to manually figure it out! I thought Docusaurus warns about wrong internal links during building hm, but I guess it doesn't or it didn't for this one for some reason?

sodic replied:
Docosaurus throws an error, yes.

Why do you think it doesn't warn about it?

FranjoMindek replied:
I changed this link because I renamed the file from backends to databases. I've added rest of them now.

Martinsos replied:
Ah I didn't realize you renamed the file! That is why I asked about why we didn't get the warning before, because I thought it is an old broken link that you picked up and fixed. Ok got it.

Martinsos replied:
Feel free to resolve this comment once you read it.

Martinsos commented on web/static/_redirects:
```diff
@@ -8,6 +8,7 @@
 /docs/deploying                    /docs/deployment/intro                    301
 /docs/advanced/deployment/overview /docs/deployment/intro                    301
 /docs/integrations/css-frameworks  /docs/project/css-frameworks             301
+/docs/data-model/backends          /docs/data-model/databases               301
```
Hm but this is probably not 1 on 1 mapping? I am guessing that backends was wider than just databases? If it is just name change then makews sense, but if not and backends was wider, maybe better not to map at all?

Martinsos reviewed (commented):
Looks good to me! I will leave approving to @infomiho though, since he knows the Auth part better.

sodic commented on web/docs/auth/overview.md:
```diff
@@ -313,7 +313,7 @@ export function Main() {
 </Tabs>
 
 :::tip
-Since the `user` prop is only available in a page's React component: use the `user` prop in the page's React component and the `useAuth` hook in any other React component.
+While the `user` prop is only available in an authenticated page's React component, it is the preferred option because we can be certain that the user exists on that page. On the other hand, `useAuth` can be used when the user might not exist, making it better suited for non-authenticated components.
 :::
```
I'd move this to the beginning of the chapter. It's best to warn users about a better approach at the beginning.

And we can lose the tip - the text is part of the document's theme.

sodic commented on web/docs/auth/overview.md:
```diff
@@ -184,7 +184,7 @@ There are two ways to access the `user` object on the client:
 #### Using the `user` prop
 
-If the page's declaration sets `authRequired` to `true`, the page's React component receives the `user` object as a prop:
+The preferred way to fetch user data on authenticated pages is by using the `user` prop. If the page's declaration sets `authRequired` to `true`, the page's React component receives the `user` object as a prop:
```
```suggestion
If the page's declaration sets `authRequired` to `true`, the page's React component receives the `user` object as a prop. This is the simplest way to access the user inside an authenticated page:
```

Martinsos commented:
@FranjoMindek did I get something wrong, or did you add quite some changes after my last review -> what happened there? Are those still in the initial scope of this PR, or are these additional fixes?

FranjoMindek commented:
Initial commit didn't capture all places referencing the old name of the document. I've added rest of them.
Filip also wanted to word things a bit differently so I've added those too.
Lastly, we found mentions of "wasp-lang" instead of wasp in some old document so Filip said I should fix those too.

FranjoMindek commented on waspc/src/Wasp/Generator/DbGenerator.hs:
```diff
@@ -258,7 +258,7 @@ generatePrismaClient spec projectRootDir = do
     generatePrismaClientIfEntitiesExist
       | entitiesExist =
-          either (Just . GenericGeneratorError) (const Nothing) <$> DbOps.generatePrismaClient projectRootDir
+        either (Just . GenericGeneratorError) (const Nothing) <$> DbOps.generatePrismaClient projectRootDir
```
There is mention of it in the docs, I think I've set it up but I'll check it out.

infomiho commented on web/docs/auth/overview.md:
````diff
@@ -1173,7 +1170,7 @@ The extra fields can be either **objects** or **render functions** (you can comb
    The render function has the following signature:
 
-   ```ts
+   ```
````
Maybe we can improve this code to be valid TS so we get syntax highlighting? That could be a nice improvement for readers.

FranjoMindek replied:
What would you write? 
If we want syntax highlighting we need something like:
```ts
type renderFn = (form: UseFormReturn, state: FormState) => React.ReactNode
```

But this may lead on readers that such type exists.

infomiho replied:
I'd write it like that:
```ts
type RenderFn = (form: UseFormReturn, state: FormState) => React.ReactNode
```
We are not writing an import statement, I think the readers would get it 👍

FranjoMindek replied:
Now that I've looked into the source code, we have this:
```ts
// PRIVATE API
export type AdditionalSignupFieldRenderFn = (
  hookForm: UseFormReturn<LoginSignupFormFields>,
  formState: FormState
) => React.ReactNode
```

Even though this is a private API I still think it would be preferable to include the same as the source code. It could be helpful to possible future contributors or anyone who wants to dig deeper into wasp. I would omit the `// PRIVATE API` comment and `export` for the docs.
@sodic you had some reservations around this, do you think this is fine?

sodic replied:
Hm, maybe too much for my taste. I'd go with what you had originally (just the type signature) and maybe even inline it into the text.

That said, I am fine with whatever you like best.

One final piece of advice, ask yourself this: _"If I could have proper syntax highlighting (or none) for all options, which one would I pick?"_ And then go with that one.

We always want the best possible content, not the content that makes our tools the happiest :)

Anyway, pick whatever you like and feel free to resolve!

sodic commented on web/docs/auth/social-auth/overview.md:
```diff
@@ -215,7 +215,8 @@ export const userSignupFields = defineUserSignupFields({
 
-You can query the user's `isSignupComplete` flag on the client with the [`useAuth()`](../../auth/overview) hook.
+You can check the `isSignupComplete` flag on the `user` object.
+If you're on an authenticated page the `user` can be fetched through the [props](../../auth/overview#getting-the-user-in-authenticated-routes), otherwise use the [`useAuth()`](../../auth/overview#getting-the-user-in-non-authenticated-routes) hook instead.
```
I'd change it to something like this:
```suggestion
Authenticated pages come with the [`user` prop](../../auth/overview#getting-the-user-in-authenticated-routes) which gives you access to the current user. If the `user` prop is out of reach, fetch the current user with the  [`useAuth()` hook](../../auth/overview#getting-the-user-in-non-authenticated-routes).
```
Explanation:
- Active voice is better than passive voice
- Shorter sentences are easier to parse
- "fetched" implies a network request, which the prop doesn't do.
- URLS look more natural when they span the entire noun phrase (e.g., "useAuth hook" instead of "useAuth").
- Other stuff just has to do with the flow and is probably a little subjective.

I also liked your previous explanation and we seemed to have lost it. You can append it here:
```
As a rule of thumb:
- If a component only makes sense for authenticated users, it's best to rely on the `user` prop.
- If a component works for both authenticated and non-authenticated users and needs information about the current user, reach for the `useAuth` hook.

sodic replied:
Also, seems like we repeat the same thing (hook vs prop) both here (in social auth overview) and in the regular auth overview.

@infomiho Is this on purpose or did we miss it?

sodic replied:
Yeah, it's not ideal. Because it's not really about routes/pages. It's about any component inside the tree:
- Anything that's in the tree whose root is an authenticated page should use the prop (and we should pass it down accordingly)
- Other stuff should use `useAuth`

As I was writing the first point, something seemed off. @FranjoMindek What would be the best way to do this, if we had a clean slate? Would the user always sit in the context (instead of getting passed via the props)?

On the topic on where to put this rule of thumb... If you can't fit it anywhere, we can leave it out. Like you said, we already have those titles, they probably explain it well enough.

FranjoMindek replied:
> ### Getting the `user` in non-authenticated routes
Now that I think about it, it may lead people to think that it can't be used in authenticated routes.
We do say that it's a `useQuery` wrapper, but 🤷

sodic replied:
Ok, let's merge this know and we'll come back to this discussion at a different time (has nothing to do with the PR anyway).

sodic reviewed (approved):
Nice work. I'm approving but left some suggestions
